File: src/RecipeApp.CoreApi/Features/HealthChecks/SampleHealthCheck.cs
namespace RecipeApp.CoreApi.Features.HealthChecks;

/// <summary>
/// https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/health-checks?view=aspnetcore-7.0
/// </summary>
public class SampleHealthCheck : IHealthCheck
{
    public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
    {
        var dict = new Dictionary<string, object>
            {
                { "Test1", "Test1"},
                { "Test2", "Test2"},
            };
        IReadOnlyDictionary<string, object> readonlyDict = new ReadOnlyDictionary<string, object>(dict);

        try
        {
            const bool isHealthy = true;

            if (isHealthy)
                return Task.FromResult(HealthCheckResult.Healthy("SampleHealthCheck is healthy.", readonlyDict));


            return Task.FromResult(HealthCheckResult.Unhealthy("SampleHealthCheck is Unhealthy.", null, readonlyDict));
        }
        catch (Exception ex)
        {
            return Task.FromResult(HealthCheckResult.Unhealthy("SampleHealthCheck is Unhealthy.", ex, readonlyDict));
        }
    }
}

File: src/RecipeApp.FCoreApi/Features/HealthChecks/SampleHealthCheck.fs
namespace RecipeApp.FCoreApi.Features.HealthChecks

open System.Collections.Generic
open System.Collections.ObjectModel
open System.Threading

open Microsoft.Extensions.Diagnostics.HealthChecks

type SampleHealthCheck() =
    interface IHealthCheck with
        member this.CheckHealthAsync(context: HealthCheckContext , cancellationToken: CancellationToken) =
            async {

                let dict = new Dictionary<string, System.Object>()
                dict.Add("Test1", "Test1")
                dict.Add("Test2", "Test2")

                let readonlyDict = new ReadOnlyDictionary<string, System.Object>(dict);

                try
                    let isHealthy = true

                    if isHealthy
                    then
                        return HealthCheckResult.Healthy("SampleHealthCheck is healthy.", readonlyDict)
                    else
                        return HealthCheckResult.Healthy("SampleHealthCheck is Unhealthy.")
                with
                    | ex -> return HealthCheckResult.Unhealthy("SampleHealthCheck is Unhealthy.", ex)
            }
            |> Async.StartAsTask


